File: format-latex/src/main/java/digital/vasic/yole/format/latex/LatexSyntaxHighlighter.java
package digital.vasic.yole.format.latex;

import digital.vasic.yole.frontend.textview.SyntaxHighlighterBase;
import digital.vasic.yole.model.AppSettings;
import android.text.Spannable;
import android.text.style.ForegroundColorSpan;
import java.util.regex.Pattern;
import java.util.regex.Matcher;

public class LatexSyntaxHighlighter extends SyntaxHighlighterBase {
    private static final Pattern LATEX_COMMANDS = Pattern.compile("\\\\[a-zA-Z]+(\\{[^}]*\\})*");
    private static final Pattern LATEX_INLINE_MATH = Pattern.compile("\\$[^$]+\\$");
    private static final Pattern LATEX_DISPLAY_MATH = Pattern.compile("\\$\\$.*?\\$\\$");
    private static final Pattern LATEX_COMMENTS = Pattern.compile("%.*$");

    public LatexSyntaxHighlighter(AppSettings appSettings) {
        super(appSettings);
    }

    @Override
    public void applySyntaxHighlighting(Spannable spannable, int color) {
        if (spannable == null) return;

        String text = spannable.toString();

        // Highlight LaTeX commands
        Matcher commandMatcher = LATEX_COMMANDS.matcher(text);
        while (commandMatcher.find()) {
            spannable.setSpan(
                new ForegroundColorSpan(getHighlightColor()),
                commandMatcher.start(),
                commandMatcher.end(),
                Spannable.SPAN_EXCLUSIVE_EXCLUSIVE
            );
        }

        // Highlight inline math
        Matcher inlineMathMatcher = LATEX_INLINE_MATH.matcher(text);
        while (inlineMathMatcher.find()) {
            spannable.setSpan(
                new ForegroundColorSpan(getMathColor()),
                inlineMathMatcher.start(),
                inlineMathMatcher.end(),
                Spannable.SPAN_EXCLUSIVE_EXCLUSIVE
            );
        }

        // Highlight display math
        Matcher displayMathMatcher = LATEX_DISPLAY_MATH.matcher(text);
        while (displayMathMatcher.find()) {
            spannable.setSpan(
                new ForegroundColorSpan(getMathColor()),
                displayMathMatcher.start(),
                displayMathMatcher.end(),
                Spannable.SPAN_EXCLUSIVE_EXCLUSIVE
            );
        }

        // Highlight comments
        Matcher commentMatcher = LATEX_COMMENTS.matcher(text);
        while (commentMatcher.find()) {
            spannable.setSpan(
                new ForegroundColorSpan(getCommentColor()),
                commentMatcher.start(),
                commentMatcher.end(),
                Spannable.SPAN_EXCLUSIVE_EXCLUSIVE
            );
        }
    }

    private int getHighlightColor() {
        return 0xFF0066CC; // Blue for commands
    }

    private int getMathColor() {
        return 0xFF009900; // Green for math
    }

    private int getCommentColor() {
        return 0xFF666666; // Gray for comments
    }
}

File: app/src/main/kotlin/digital/vasic/yole/format/latex/LatexSyntaxHighlighter.kt
/*#######################################################
 *
 *   Maintained 2018-2025 by Gregor Santner <gsantner AT mailbox DOT org>
 *   License of this file: Apache 2.0
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 *########################################################*/
package digital.vasic.yole.format.latex

import android.graphics.Paint
import digital.vasic.yole.frontend.textview.SyntaxHighlighterBase
import digital.vasic.yole.model.AppSettings
import java.util.regex.Pattern

class LatexSyntaxHighlighter(appSettings: AppSettings) : SyntaxHighlighterBase(appSettings) {

    override fun configure(paint: Paint?): SyntaxHighlighterBase {
        return super.configure(paint)
    }

    override fun generateSpans() {
        createTabSpans(_tabSize)

        // Highlight LaTeX commands
        createColorSpanForMatches(LATEX_COMMANDS, HIGHLIGHT_COLOR)

        // Highlight inline math
        createColorSpanForMatches(LATEX_INLINE_MATH, MATH_COLOR)

        // Highlight display math
        createColorSpanForMatches(LATEX_DISPLAY_MATH, MATH_COLOR)

        // Highlight comments
        createColorSpanForMatches(LATEX_COMMENTS, COMMENT_COLOR)

        // Highlight environments
        createColorSpanForMatches(LATEX_ENVIRONMENTS, ENVIRONMENT_COLOR)
    }

    companion object {
        @JvmField
        val LATEX_COMMANDS: Pattern = Pattern.compile("\\\\[a-zA-Z]+(\\{[^}]*\\})*")

        @JvmField
        val LATEX_INLINE_MATH: Pattern = Pattern.compile("\\$[^$]+\\$")

        @JvmField
        val LATEX_DISPLAY_MATH: Pattern = Pattern.compile("\\$\\$.*?\\$\\$")

        @JvmField
        val LATEX_COMMENTS: Pattern = Pattern.compile("%.*$")

        @JvmField
        val LATEX_ENVIRONMENTS: Pattern = Pattern.compile("\\\\(begin|end)\\{[^}]+\\}")

        private const val HIGHLIGHT_COLOR = 0xFF0066CC.toInt()  // Blue for commands
        private const val MATH_COLOR = 0xFF009900.toInt()        // Green for math
        private const val COMMENT_COLOR = 0xFF666666.toInt()     // Gray for comments
        private const val ENVIRONMENT_COLOR = 0xFFCC6600.toInt() // Orange for environments
    }
}
